### ISSUE TITLE
Warnings in utf8 subcategories do nothing in isolation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11971]
Migrated from [rt.perl.org#111338](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=111338) (status was 'resolved')

Searchable as RT111338$

**Comment by p5pRT** [comment]
### From @arc

#### Created by @arc



When used in isolation\, the C\<non\_unicode>\, C\<nonchar>\, and
C\<surrogate> warning subcategories don't trigger the emission of
the relevant warnings; though using their parent category C\<utf8>
does\.

For example&#8203;:

$ \./perl \-CS \-e 'use warnings "nonchar"; print chr 0xffff' > /dev/null
$ \./perl \-CS \-e 'use warnings "utf8"; print chr 0xffff' > /dev/null
Unicode non\-character U\+FFFF is illegal for open interchange at \-e line 1\.

Patch attached\, containing failing tests\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl 5.15.8:

Configured by aaron at Sat Feb 25 12:07:10 GMT 2012.

Summary of my perl5 (revision 5 version 15 subversion 8) configuration:
  Commit id: 236cbe8d7a6747fb6144c93dd767895f8acffd00
  Platform:
    osname=linux, osvers=3.0.0-14-server, archname=x86_64-linux
    uname='linux redacted 3.0.0-14-server #23~lucid1-ubuntu smp thu
dec 8 14:50:26 utc 201
1 x86_64 gnulinux '
    config_args='-des -Dusedevel -Dprefix=/var/tmp/bleadperl'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib
/lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.1.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.15.8:
    /var/tmp/bleadperl/lib/site_perl/5.15.8/x86_64-linux
    /var/tmp/bleadperl/lib/site_perl/5.15.8
    /var/tmp/bleadperl/lib/5.15.8/x86_64-linux
    /var/tmp/bleadperl/lib/5.15.8
    .


Environment for perl 5.15.8:
    HOME=/home/aaron
    LANG=en_GB.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

-- 
Aaron Crane ** http://aaroncrane.co.uk/

```
</details>

**Comment by p5pRT** [comment]
### From @arc

<details><summary>0001-Failing-tests-for-warnings-in-utf8-subcategories.patch</summary>

```diff
From 351623ab2a8483c12d18ca986d23066ccf822dd8 Mon Sep 17 00:00:00 2001
From: Aaron Crane <arc@cpan.org>
Date: Sat, 25 Feb 2012 13:40:11 +0000
Subject: [PATCH] Failing tests for warnings in utf8 subcategories

---
 t/lib/warnings/utf8 |   30 ++++++++++++++++++++++++++++++
 1 files changed, 30 insertions(+), 0 deletions(-)

diff --git a/t/lib/warnings/utf8 b/t/lib/warnings/utf8
index 8fb96c9..44e3c4e 100644
--- a/t/lib/warnings/utf8
+++ b/t/lib/warnings/utf8
@@ -349,6 +349,36 @@ EXPECT
 Unicode surrogate U+D800 is illegal in UTF-8 at - line 6.
 Unicode non-character U+FFFF is illegal for open interchange at - line 7.
 ########
+# NAME C<use warnings "nonchar"> works in isolation
+require "../test.pl";
+use warnings 'nonchar';
+my $file = tempfile();
+open(my $fh, "+>:utf8", $file);
+print $fh "\x{FFFF}", "\n";
+close $fh;
+EXPECT
+Unicode non-character U+FFFF is illegal for open interchange at - line 5.
+########
+# NAME C<use warnings "surrogate"> works in isolation
+require "../test.pl";
+use warnings 'surrogate';
+my $file = tempfile();
+open(my $fh, "+>:utf8", $file);
+print $fh "\x{D800}", "\n";
+close $fh;
+EXPECT
+Unicode surrogate U+D800 is illegal in UTF-8 at - line 5.
+########
+# NAME C<use warnings "non_unicode"> works in isolation
+require "../test.pl";
+use warnings 'non_unicode';
+my $file = tempfile();
+open(my $fh, "+>:utf8", $file);
+print $fh "\x{110000}", "\n";
+close $fh;
+EXPECT
+Code point 0x110000 is not Unicode, may not be portable at - line 5.
+########
 require "../test.pl";
 no warnings 'utf8';
 my $file = tempfile();
-- 
1.7.0.4


```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

I applied your patch to blead\, commit
920e47bbc431efd5c1ca9d8b85fb88c172326b14

but changing the tests to TODOs to avoid them failing blead

Are there objections to fixing this in 5\.16?  It is to just add these
warning categories to two ck\_warn calls\.
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From tchrist@perl.com

>I applied your patch to blead\, commit
>920e47bbc431efd5c1ca9d8b85fb88c172326b14

>but changing the tests to TODOs to avoid them failing blead

>Are there objections to fixing this in 5\.16?  It is to just add these
>warning categories to two ck\_warn calls\.

I don't get a vote\, but I \*would\* like to see this fixed\.

Thanks\,

\-\-tom

**Comment by p5pRT** [comment]
### From @rjbs

\* Karl Williamson via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-03\-04T14&#8203;:28&#8203;:54\]
> Are there objections to fixing this in 5\.16?  It is to just add these
> warning categories to two ck\_warn calls\.

Let's get the fix in a smoke\-me branch for now\, and we can decide whether it
goes in 5\.16\.0 or 5\.16\.1 in short order\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @rjbs

\* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-03\-05T10&#8203;:29&#8203;:23\]
> \* Karl Williamson via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-03\-04T14&#8203;:28&#8203;:54\]
> > Are there objections to fixing this in 5\.16?  It is to just add these
> > warning categories to two ck\_warn calls\.
> 
> Let's get the fix in a smoke\-me branch for now\, and we can decide whether it
> goes in 5\.16\.0 or 5\.16\.1 in short order\.

Good grief\, how did I lose track of this for so long?

Karl\, how did this smoking go?  I thought we'd applied this\, but the ticket is
still open\.  I don't see an obvious\-named smoke\-me branch\, though\.

Fill me in and we'll get it sorted out ASAP\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/16/2012 09&#8203;:07 PM\, Ricardo Signes wrote&#8203;:
> \* Ricardo Signes\<perl\.p5p@&#8203;rjbs\.manxome\.org>  \[2012\-03\-05T10&#8203;:29&#8203;:23\]
>> \* Karl Williamson via RT\<perlbug\-followup@&#8203;perl\.org>  \[2012\-03\-04T14&#8203;:28&#8203;:54\]
>>> Are there objections to fixing this in 5\.16?  It is to just add these
>>> warning categories to two ck\_warn calls\.
>>
>> Let's get the fix in a smoke\-me branch for now\, and we can decide whether it
>> goes in 5\.16\.0 or 5\.16\.1 in short order\.
>
> Good grief\, how did I lose track of this for so long?
>
> Karl\, how did this smoking go?  I thought we'd applied this\, but the ticket is
> still open\.  I don't see an obvious\-named smoke\-me branch\, though\.

Just the typical failures; It was a non\-obvious branch name&#8203;:

http&#8203;://perl\.develop\-help\.com/?b=smoke\-me%2Fkhw\-regex
This also smoked \#111560

I am ready to apply one or both\.

**Comment by p5pRT** [comment]
### From @rjbs

\* Karl Williamson \<public@&#8203;khwilliamson\.com> \[2012\-04\-17T11&#8203;:06&#8203;:42\]
> Just the typical failures; It was a non\-obvious branch name&#8203;:
> 
> http&#8203;://perl\.develop\-help\.com/?b=smoke\-me%2Fkhw\-regex
> This also smoked \#111560
> 
> I am ready to apply one or both\.

Thanks\, please go ahead\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/17/2012 09&#8203;:06 AM\, Karl Williamson wrote&#8203;:
> On 04/16/2012 09&#8203;:07 PM\, Ricardo Signes wrote&#8203;:
>> \* Ricardo Signes\<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-03\-05T10&#8203;:29&#8203;:23\]
>>> \* Karl Williamson via RT\<perlbug\-followup@&#8203;perl\.org>
>>> \[2012\-03\-04T14&#8203;:28&#8203;:54\]
>>>> Are there objections to fixing this in 5\.16? It is to just add these
>>>> warning categories to two ck\_warn calls\.
>>>
>>> Let's get the fix in a smoke\-me branch for now\, and we can decide
>>> whether it
>>> goes in 5\.16\.0 or 5\.16\.1 in short order\.
>>
>> Good grief\, how did I lose track of this for so long?
>>
>> Karl\, how did this smoking go? I thought we'd applied this\, but the
>> ticket is
>> still open\. I don't see an obvious\-named smoke\-me branch\, though\.
>
> Just the typical failures; It was a non\-obvious branch name&#8203;:
>
> http&#8203;://perl\.develop\-help\.com/?b=smoke\-me%2Fkhw\-regex
> This also smoked \#111560
>
> I am ready to apply one or both\.
>

Now done\, as a099aed4f4cbb1c9ea2766707caa76d250e3b25a

I forgot that 111560 had already been applied

**Comment by p5pRT** [comment]
### From @khwilliamson

Fixed by a099aed4f4cbb1c9ea2766707caa76d250e3b25a

 \-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
